pkg/resource/deploy/snapshot.go

```diff
@@ -49,6 +49,130 @@ func NewSnapshot(manifest Manifest, secretsManager secrets.Manager,
 	}
 }
 
+// Topsort sorts this snapshot so that it is topologically sorted with respect to dependencies (where a dependency could
+// be a provider, parent-child relationship, dependency, and so on). Resources in the resulting snapshot will appear in
+// an order such that all dependencies of a resource will appear before the resource itself.
+//
+// This method is generally only used for repairing invalid snapshots, since most snapshots are built in response to
+// resource registrations from a program, and programs are required to submit such registrations in a
+// dependency-respecting order. Note that sortedness is a necessary but not sufficient condition for a snapshot to be
+// valid; the VerifyIntegrity method should be used to ensure that a snapshot is well-formed.
+func (snap *Snapshot) Topsort() error {
+	sorted := []*resource.State{}
+
+	// We implement the sort using a post-order depth-first search, keeping track of nodes we have visited and terminating
+	// when we have seen them all. It is not possible to sort a snapshot with cycles (and indeed, such snapshots will
+	// never be valid Pulumi states). To this end we also keep track of the path we are currently visiting so that we can
+	// spot if we are in a cycle.
+	visiting := map[*resource.State]bool{}
+	visited := map[*resource.State]bool{}
+
+	// When traversing dependencies, we'll need to look them up by URN. It is possible that the same URN exists twice in a
+	// snapshot: in the case that the snapshot represents the state mid-way through a replacement, both the old and new
+	// resources could appear in the snapshot. Dependencies between old and new resources are permitted, so it's important
+	// that we know which is which and don't disambiguate by URN alone. To this end we keep track of two lookup tables --
+	// old resources (identifiable by their Delete flag being set) and new resources.
+	oldsByURN := map[resource.URN]*resource.State{}
+	newsByURN := map[resource.URN]*resource.State{}
+	for _, state := range snap.Resources {
+		if state.Delete {
```

Comment:
This might not be enough. You can potentially have _multiple_ resources left over pending delete :(

Reply:
Discussed offline: added some test cases to demonstrate things that we can handle and disclaimers to the documentation for the function to try and and communicate that in general, it's not always possible for us to sort snaphots where there are ambiguous URNs.